tiledb/common/thread_pool/producer_consumer_queue.h

```diff
@@ -0,0 +1,232 @@
+   * @param item Item to be pushed onto the queue.
+   * @return bool indicating whether the item was successfully pushed or not.
+   */
+  auto push(const Item& item) {
+    std::unique_lock lock{mutex_};
+    if (closed_) {
+      return false;
+    }
+
+    if constexpr (std::is_same<Container, std::queue<Item>>::value) {
+      queue_.push(item);
+    } else if constexpr (std::is_same<Container, std::deque<Item>>::value) {
+      queue_.push_front(item);
+    } else {
+      queue_.no_push(item);
+    }
+
+    // lock.unlock();
+    cv_.notify_one();
+    return true;
+  }
+
+
+  /**
+   * Push an item onto the producer-consumer queue.  This producer-consumer
+   * queue is unbounded; there is no risk of the caller being put to sleep.  If
+   * the queue is closed, the item is not pushed and false is returned.
+   *
+   * @param item Item to be pushed onto the queue.
+   * @return bool indicating whether the item was successfully pushed or not.
+   */
+  auto push(Item&& item) {
+    std::unique_lock lock{mutex_};
+    if (closed_) {
+      return false;
+    }
+
+    if constexpr (std::is_same<Container, std::queue<Item>>::value) {
+      queue_.push(std::move(item));
+    } else if constexpr (std::is_same<Container, std::deque<Item>>::value) {
+      queue_.push_front(std::move(item));
+    } else {
+      queue_.no_push(item);
+    }
+
+    // lock.unlock();
+    cv_.notify_one();
+    return true;
+  }
+
+  /**
+   * Try to pop an item from the queue.  If no item is available
+   * the function returns nothing.  It does not sleep.
+   *
+   * @returns Item from the queue, if available, otherwise nothing.
+   */
+  std::optional<Item> try_pop() {
+    std::scoped_lock lock{mutex_};
+
+    if (queue_.empty() || closed_) {
+      return {};
+    }
+    Item item = queue_.front();
+
+    if constexpr (std::is_same<Container, std::queue<Item>>::value) {
+      queue_.pop(item);
+    } else if constexpr (std::is_same<Container, std::deque<Item>>::value) {
+      queue_.pop_front();
+    } else {
+      queue_.no_pop(item);
+    }
+
+    return item;
+  }
+
+  /**
+   * Pop an item from the queue.  If the queue is empty, the calling
+   * thread will wait on a condition variable until an item becomes
+   * available.  If the queue is empty and the queue is closed 
+   * (shutting down), nothing is returned.  If the queue is not
+   * empty and the queue is closed, an item will be returned.
+   *
+   * @returns Item from the queue, if available, otherwise nothing.
+   */
+  std::optional<Item> pop() {
+    std::unique_lock lock{mutex_};
+
+    cv_.wait(lock, [this]() { return closed_ || !queue_.empty(); });
+
+    if (closed_ && queue_.empty()) {
+      return {};
+    }
+    Item item = queue_.front();
+
+    if constexpr (std::is_same<Container, std::queue<Item>>::value) {
+      queue_.pop(item);
+    } else if constexpr (std::is_same<Container, std::deque<Item>>::value) {
+      queue_.pop_front();
+    } else {
+      queue_.no_pop(item);
```

Comment:
No types define this method -- the idea is to fail compilation if neither a queue nor a deque is attempted to be used.  We could use a static assert with a more meaningful message (rather than encoding it obliquely in the function name).

Reply:
Yes please, or at least a comment. I was slightly confused by the intent here 😄